src/Sarif.Multitool.Library/Rules/SARIF2016.FileUrisShouldBeRelative.cs

```diff
@@ -72,14 +73,32 @@ private void AnalyzeResultLocations(IList<Location> locations, string locationsP
 
                 if (location?.PhysicalLocation?.ArtifactLocation?.Uri != null)
                 {
-                    if (location.PhysicalLocation.ArtifactLocation.Uri.IsAbsoluteUri && location.PhysicalLocation.ArtifactLocation.Uri.IsFile)
+                    if (location.PhysicalLocation.ArtifactLocation.Uri.IsAbsoluteUri)
                     {
-                        // {0}: The file location '{1}' is specified with absolute URI. Prefer a relative
-                        // reference together with a uriBaseId property.
-                        LogResult(
-                            locationPointer.AtProperty(SarifPropertyName.PhysicalLocation).AtProperty(SarifPropertyName.ArtifactLocation).AtProperty(SarifPropertyName.Uri),
-                            nameof(RuleResources.SARIF2016_FileUrisShouldBeRelative_Note_Default_Text),
-                            location.PhysicalLocation.ArtifactLocation.Uri.OriginalString);
+                        if (location.PhysicalLocation.ArtifactLocation.Uri.IsFile)
+                        {
+                            // {0}: The file location '{1}' is specified with absolute URI. Prefer a relative
+                            // reference together with a uriBaseId property.
+                            LogResult(
+                                locationPointer.AtProperty(SarifPropertyName.PhysicalLocation).AtProperty(SarifPropertyName.ArtifactLocation).AtProperty(SarifPropertyName.Uri),
+                                nameof(RuleResources.SARIF2016_FileUrisShouldBeRelative_Note_Default_Text),
+                                location.PhysicalLocation.ArtifactLocation.Uri.OriginalString);
+                        }
+                    }
+                    else
+                    {
+                        if (location.PhysicalLocation.ArtifactLocation.Uri.OriginalString.StartsWith("/"))
```

Comment:
>location.PhysicalLocation.ArtifactLocation.Uri.OriginalString [](start = 28, length = 61)

Are we sure this is right? Try creating a relative URL like so, new Url(@"\myTest\myFile.cpp", UriKind.Relative). I think URL.ToString() will convert to forward slashes but OriginalString will have back slashes.

Reply:
I did some testing:
```csharp
            var uri = new Uri(@"\myTest\myFile.cpp", UriKind.Relative);
            var uri2 = new Uri(@"/myTest/myFile.cpp", UriKind.Relative);

            Console.WriteLine(uri.OriginalString);
            Console.WriteLine(uri.ToString());

            Console.WriteLine(uri2.OriginalString);
            Console.WriteLine(uri2.ToString());
```

the value printed:
```
\myTest\myFile.cpp
\myTest\myFile.cpp
/myTest/myFile.cpp
/myTest/myFile.cpp
```

Based on this, I updated the logic to check for `\` and for `/`.

What do you think?

---
In reply to: [533633936](https://github.com/microsoft/sarif-sdk/pull/2167#discussion_r533633936) [](ancestors = 533633936)

Reply:
Well, I think this raises the possibility of updating this analysis further! Confusion around forward or back slashes is really prevalent for file URLs, particularly when considering Windows. We should probably encourage the use of forward slashes only (and update the validator to check for this). Try this test: new Uri(@"c:\test\myfile.cpp", UriKind.Absolute). Does uri.ToString() convert back slashes to forward in this case?

---
In reply to: [533669651](https://github.com/microsoft/sarif-sdk/pull/2167#discussion_r533669651) [](ancestors = 533669651,533633936)

Reply:
Below the results:
```
c:\test\myfile.cpp (OriginalString)
[file:///c:/test/myfile.cpp](file:///c:/test/myfile.cpp) (tostring())
```

---
In reply to: [533692935](https://github.com/microsoft/sarif-sdk/pull/2167#discussion_r533692935) [](ancestors = 533692935,533669651,533633936)

Reply:
So, I ran a test and is a relative URI contains back slashes, they are preserved in the final constructed URI. This is not good, as this value could be compared to other (say, locally constructed) URIs. So let's do update this analysis to prevent back slashes in any relative URI. You can open a new issue and do it as follow on work, if you like.

[file:///c:/mydir\anotherDir\myFile](file:///c:/mydir\anotherDir\myFile)

---
In reply to: [533694852](https://github.com/microsoft/sarif-sdk/pull/2167#discussion_r533694852) [](ancestors = 533694852,533692935,533669651,533633936)

Reply:
ok, so, we would have:
1. if relative and start with "/" -> that is a problem
2. if relative and contains "\" -> that is a problem

Is that right?

---
In reply to: [533731429](https://github.com/microsoft/sarif-sdk/pull/2167#discussion_r533731429) [](ancestors = 533731429,533694852,533692935,533669651,533633936)

Reply:
Just updated the logic. Can you validate and check the new message that i just created?
Thank you!

---
In reply to: [533735403](https://github.com/microsoft/sarif-sdk/pull/2167#discussion_r533735403) [](ancestors = 533735403,533731429,533694852,533692935,533669651,533633936)